### ISSUE TITLE
allow specifying "annotation only" injection in a namespace

### ISSUE THREAD

**Issue by jonmoter** [istio/istio#14087]
**Describe the feature request**

I have a cluster where I have some namespaces that I want to always inject a sidecar.

I also have one or more namespaces where I want to selectively enable injection via a pod annotation.  We have a lot of services in our `default` namespace, and I want to enable the sidecars for a few at a time, rather than all at once.

So I want the logic to look kindof like:

```
nsInject = namespace.labels['istio-injection']
podInject = pod.annotations['sidecar.istio.io/inject']

shouldInject = (nsInject == 'enabled') || (nsInject == 'annotation' && podInject == 'true')
```

**Describe alternatives you've considered**

If I don't set the `istio-injection=enabled` label on the namespace, the admission controller logic doesn't event get executed.  If my default policy is `enabled`, I'd have to blacklist every podspec in the `default` namespace that I don't want it injected in.

I looked at the [exception docs](https://istio.io/docs/setup/kubernetes/additional-setup/sidecar-injection/#more-control-adding-exceptions) and considered adding a `neverInjectSelector` and/or a `alwaysInjectSelector`, but those only take the pod labels into account. I can't add logic that looks at the namespace labels.

**Additional context**

The context is I'm coming up with a strategy for rolling out Istio in our production Kubernetes clusters.  I want to do it gradually, but I really like the convenience of being able to set the policy for an entire namespace.

Another option would be some mechanism of setting the policy on a namespace level.  So the `MutatingWebhookConfiguration` would be configured so that it would work on any namespace where `istio-injection` was set, and the value of `enabled` or `disabled` would be the policy for that namespace.

**Comment by Stono** [comment]
I don't know if this helps, but we slightly invert the logic of the injector.

We set:
```
  namespaceSelector:
    matchExpressions:
    - key: istio-injection
      operator: NotIn
      values:
      - disabled
```

And then we set our policy to be `disabled` too.

We then label certain namespaces with `istio-injection: disabled`:
```
❯ k get ns -L istio-injection | grep disabled
core-system                             Active   358d    disabled
istio-system                            Active   346d    disabled
kube-system                             Active   364d    disabled
```

This means by default; all namespaces get processed by the webhook and we don't have to worry about labeling them, but only deployments with the `sidecar.istio.io/inject: "true"` annotation get a sidecar.

**Comment by Stono** [comment]
It actually looks like following my [previous issue](https://github.com/istio/istio/issues/5576) about that a flag was added to enable it: https://github.com/istio/istio/pull/6704/files#diff-297a80504fd255768ce50ed37f642ffeR26

**Comment by oresot** [comment]
From the docs [here](https://istio.io/docs/setup/kubernetes/additional-setup/sidecar-injection/#policy) I understand that the `sidecar.istio.io/inject: "true"` annotation will **only** work when the `istio-sidecar-injector` `policy` is set to `disabled`.

I would expect that when `policy` is `enabled`, as it is by default, it would inject the sidecars when:

- Namespace is annotated with inject=true -> Injects sidecar in all pods that doesn't explicitly disable injection _(this already happens)_
or
- Namespace is not annotated or annotated with inject=false  -> Injects sidecar only in pods that explicitly enable injection

What is the reason behind this limitation? Does it really makes sense how it is implemented right now?

**Comment by howardjohn** [comment]
I am not sure this is feasible to do currently, we can only decide send request/don't send request to the sidecar injector, and then the injector only gets the pod spec, so we cannot see if the namespace is annotated with "annotation" only.

However, in k8s 1.15 they added some new stuff to webhooks, and I think it could be possible with objectSelector -- see https://github.com/istio/istio/pull/16539

**Comment by howardjohn** [comment]
I don't think there is anything to do here. We have some workarounds presented, and a proper long term fix in #16539
